rust/foxglove/src/mcap_writer/mcap_stream.rs

```diff
@@ -0,0 +1,298 @@
+use std::{
+    fmt::Debug,
+    io::{self, Seek, SeekFrom, Write},
+    pin::Pin,
+    sync::Arc,
+    task::Poll,
+};
+
+use tokio::sync::mpsc::{error::SendError, Receiver as TokioReceiver, Sender as TokioSender};
+
+use bytes::{Bytes, BytesMut};
+use futures::{ready, Stream};
+use parking_lot::Mutex;
+
+use crate::{Context, FoxgloveError, McapWriteOptions, McapWriter, McapWriterHandle};
+
+#[derive(Default)]
+struct Inner {
+    buffer: BytesMut,
+    position: u64,
+}
+
+#[derive(Default, Clone)]
+struct SharedBuffer(Arc<Mutex<Inner>>);
+
+impl Debug for SharedBuffer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("SharedBuffer").finish_non_exhaustive()
+    }
+}
+
+impl Write for SharedBuffer {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let mut inner = self.0.lock();
+        inner.buffer.extend_from_slice(buf);
+        inner.position += buf.len() as u64;
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Seek for SharedBuffer {
+    fn seek(&mut self, pos: SeekFrom) -> std::io::Result<u64> {
+        let inner = self.0.lock();
+        match pos {
+            SeekFrom::Start(n) if inner.position == n => Ok(n),
+            SeekFrom::Current(0) => Ok(inner.position),
+            _ => Err(std::io::Error::other("seek on unseekable file")),
```

Comment:
We could support End(i64) here, for negative values. Maybe an error or noop for seeking beyond end of the buffer.